share/man/man8/ttymon.8
'\" te
.\" Copyright 1989 AT&T
.\" Copyright (C) 2004, Sun Microsystems, Inc.
.\" All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH TTYMON 8 "Feb 22, 2005"
.SH NAME
ttymon \- port monitor for terminal ports
.SH SYNOPSIS
.LP
.nf
\fB/usr/lib/saf/ttymon\fR
.fi

.LP
.nf
\fB/usr/lib/saf/ttymon\fR \fB-g\fR [\fB-d\fR \fIdevice\fR] [\fB-h\fR] [\fB-t\fR \fItimeout\fR]
     [\fB-l\fR \fIttylabel\fR] [\fB-p\fR \fIprompt\fR] [\fB-m\fR \fImodules\fR] [\fB-T\fR \fItermtype\fR]
.fi

.SH DESCRIPTION
.sp
.LP
\fBttymon\fR is a STREAMS-based TTY port monitor. Its function is to monitor
ports, to set terminal modes, baud rates, and line disciplines for the ports,
and to connect users or applications to services associated with the ports.
Normally, \fBttymon\fR is configured to run under the Service Access
Controller, \fBsac\fR(8), as part of the Service Access Facility (SAF). It is
configured using the \fBsacadm\fR(8) command. Each instance of \fBttymon\fR
can monitor multiple ports. The ports monitored by an instance of \fBttymon\fR
are specified in the port monitor's administrative file. The administrative
file is configured using the \fBpmadm\fR(8) and \fBttyadm\fR(8) commands.
When an instance of \fBttymon\fR is invoked by the \fBsac\fR command, it starts
to monitor its ports. For each port, \fBttymon\fR first initializes the line
disciplines, if they are specified, and the speed and terminal settings. For
ports with entries in \fB/etc/logindevperm\fR, device owner, group and
permissions are set. (See \fBlogindevperm\fR(4).) The values used for
initialization are taken from the appropriate entry in the TTY settings file.
This file is maintained by the \fBsttydefs\fR(8) command. Default line
disciplines on ports are usually set up by the \fBautopush\fR(8) command of
the Autopush Facility.
.sp
.LP
\fBttymon\fR then writes the prompt and waits for user input. If the user
indicates that the speed is inappropriate by pressing the BREAK key,
\fBttymon\fR tries the next speed and writes the prompt again. When valid input
is received, \fBttymon\fR interprets the per-service configuration file for the
port, if one exists, creates a \fButmpx\fR entry if required (see
\fButmpx\fR(4)), establishes the service environment, and then invokes the
service associated with the port. Valid input consists of a string of at least
one non-newline character, terminated by a carriage return. After the service
terminates, \fBttymon\fR cleans up the \fButmpx\fR entry, if one exists, and
returns the port to its initial state.
.sp
.LP
If \fIautobaud\fR is enabled for a port, \fBttymon\fR will try to determine the
baud rate on the port automatically. Users must enter a carriage return before
\fBttymon\fR can recognize the baud rate and print the prompt. Currently, the
baud rates that can be determined by \fIautobaud\fR are \fB110\fR, \fB1200\fR,
\fB2400\fR, \fB4800\fR, and \fB9600\fR.
.sp
.LP
If a port is configured as a bidirectional port, \fBttymon\fR will allow users
to connect to a service.  If a port is bidirectional, \fBttymon\fR will wait
to read a character before it prints a prompt.
.sp
.LP
If the \fIconnect-on-carrier\fR flag is set for a port, \fBttymon\fR will
immediately invoke the port's associated service when a connection request is
received. The prompt message will not be sent.
.sp
.LP
If a port is disabled, \fBttymon\fR will not start any service on that port. If
a disabled message is specified, \fBttymon\fR will send out the disabled
message when a connection request is received. If \fBttymon\fR is disabled, all
ports under that instance of \fBttymon\fR will also be disabled.
.SS "Service Invocation"
.sp
.LP
The service \fBttymon\fR invokes for a port is specified in the \fBttymon\fR
administrative file. \fBttymon\fR will scan the character string giving the
service to be invoked for this port, looking for a \fB%d\fR or a \fB%%\fR
two-character sequence. If \fB%d\fR is found, \fBttymon\fR will modify the
service command to be executed by replacing those two characters by the full
path name of this port (the device name). If \fB%%\fR is found, they will be
replaced by a single \fB%\fR. When the service is invoked, file descriptor
\fB0\fR, \fB1\fR, and \fB2\fR are opened to the port device for reading and
writing. The service is invoked with the user ID, group ID and current home
directory set to that of the user name under which the service was registered
with \fBttymon\fR. Two environment variables, \fBHOME\fR and \fBTTYPROMPT\fR,
are added to the service's environment by \fBttymon\fR. \fBHOME\fR is set to
the home directory of the user name under which the service is invoked.
\fBTTYPROMPT\fR is set to the prompt string configured for the service on the
port. This is provided so that a service invoked by \fBttymon\fR has a means of
determining if a prompt was actually issued by \fBttymon\fR and, if so, what
that prompt actually was.
.sp
.LP
See \fBttyadm\fR(8) for options that can be set for ports monitored by
\fBttymon\fR under the Service Access Controller.
.SS "System Console Invocation"
.sp
.LP
The invocation of ttymon on the system console is managed under \fBsmf\fR(5) by
the service \fBsvc:/system/console-login\fR. It provides a number of properties
within the property group \fBttymon\fR to control the invocation, as follows:
.sp
.in +2
.nf
NAME                  TYPE               TTYMON OPTION
----------------------------------------------------------
device                astring            [-d device]
nohangup              boolean            [-h]
label                 astring            [-l label]
modules               astring            [-m module1,module2]
prompt                astring            [-p prompt]
timeout               count              [-t timeout]
terminal_type         astring            [-T termtype]
.fi
.in -2
.sp

.sp
.LP
If any value is the empty string or an integer set to zero, then the option is
not passed to the ttymon invocation. The \fB-g\fR option is always specified
for this invocation. The \fB-d\fR option always defaults to \fB/dev/console\fR
if it is not set.
.sp
.LP
See \fBEXAMPLES\fR.
.SH SECURITY
.sp
.LP
\fBttymon\fR uses \fBpam\fR(3PAM) for session management. The \fBPAM\fR
configuration policy, listed through \fB/etc/pam.conf\fR, specifies the modules
to be used for \fBttymon\fR. Here is a partial \fBpam.conf\fR file with entries
for \fBttymon\fR using the UNIX session management module.
.sp
.in +2
.nf
ttymon session  required /usr/lib/security/pam_unix_session.so.1
.fi
.in -2

.sp
.LP
If there are no entries for the \fBttymon\fR service, then the entries for the
"other" service will be used.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-g\fR\fR
.ad
.RS 14n
A special invocation of \fBttymon\fR is provided with the \fB-g\fR option. This
form of the command should only be called by applications that need to set the
correct baud rate and terminal settings on a port and then connect to
\fBlogin\fR service, but that cannot be pre-configured under the SAC. The
following combinations of options can be used with \fB-g\fR:
.RE

.sp
.ne 2
.na
\fB\fB-d\fR\fIdevice\fR\fR
.ad
.RS 14n
\fIdevice\fR is the full path name of the port to which \fBttymon\fR is to
attach. If this option is not specified, file descriptor \fB0\fR must be set up
by the invoking process to a TTY port.
.RE

.sp
.ne 2
.na
\fB\fB-h\fR\fR
.ad
.RS 14n
If the -h flag is not set, \fBttymon\fR will force a hangup on the line by
setting the speed to zero before setting the speed to the default or specified
speed.
.RE

.sp
.ne 2
.na
\fB\fB-l\fR\fIttylabel\fR\fR
.ad
.RS 14n
\fIttylabel\fR is a link to a speed and TTY definition in the \fBttydefs\fR
file. This definition tells \fBttymon\fR at what speed to run initially, what
the initial TTY settings are, and what speed to try next if the user indicates
that the speed is inappropriate by pressing the BREAK key. The default speed is
9600 baud.
.RE

.sp
.ne 2
.na
\fB\fB-m\fR\fImodules\fR\fR
.ad
.RS 14n
When initializing the port, \fBttymon\fR will pop all modules on the port, and
then push \fImodules\fR in the order specified. \fImodules\fR is a
comma-separated list of pushable modules. Default modules on the ports are
usually set up by the Autopush Facility.
.RE

.sp
.ne 2
.na
\fB\fB-p\fR\fIprompt\fR\fR
.ad
.RS 14n
Allows the user to specify a prompt string. The default prompt is \fBLogin:\fR.
.RE

.sp
.ne 2
.na
\fB\fB-t\fR\fItimeout\fR\fR
.ad
.RS 14n
Specifies that \fBttymon\fR should exit if no one types anything in
\fItimeout\fR seconds after the prompt is sent.
.RE

.sp
.ne 2
.na
\fB\fB-T\fR\fItermtype\fR\fR
.ad
.RS 14n
Sets the \fBTERM\fR environment variable to \fItermtype\fR.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRSetting the Terminal Type
.sp
.LP
The following example sets the value of the terminal type (\fB-T\fR) option for
the system console \fBttymon\fR invocation:

.sp
.in +2
.nf
	svccfg -s svc:/system/console-login setprop \e
	    ttymon/terminal_type = "xterm"
	svcadm refresh svc:/system/console-login:default
.fi
.in -2
.sp

.SH ENVIRONMENT VARIABLES
.sp
.LP
If any of the \fBLC_*\fR variables ( \fBLC_CTYPE\fR, \fBLC_MESSAGES\fR,
\fBLC_TIME\fR, \fBLC_COLLATE\fR, \fBLC_NUMERIC\fR, and \fBLC_MONETARY\fR ) (see
\fBenviron\fR(5)) are not set in the environment, the operational behavior of
\fBttymon\fR for each corresponding locale category is determined by the value
of the \fBLANG\fR environment variable. If \fBLC_ALL\fR is set, its contents
are used to override both the \fBLANG\fR and the other \fBLC_*\fR variables. If
none of the above variables is set in the environment, the "C" (U.S. style)
locale determines how \fBttymon\fR behaves.
.sp
.ne 2
.na
\fB\fBLC_CTYPE\fR\fR
.ad
.RS 12n
Determines how \fBttymon\fR handles characters. When \fBLC_CTYPE\fR is set to a
valid value, \fBttymon\fR can display and handle text and filenames containing
valid characters for that locale. \fBttymon\fR can display and handle Extended
Unix Code (EUC) characters where any individual character can be 1, 2, or 3
bytes wide. \fBttymon\fR can also handle EUC characters of 1, 2, or more column
widths. In the "C" locale, only characters from ISO 8859-1 are valid.
.RE

.SH FILES
.sp
.ne 2
.na
\fB\fB/etc/logindevperm\fR\fR
.ad
.RS 21n

.RE

.sp
.LP
The command-line syntax is Stable. The SMF properties are Evolving.
.SH SEE ALSO
.sp
.LP
\fBautopush\fR(8), \fBpmadm\fR(8), \fBsac\fR(8),
\fBsacadm\fR(8), \fBsttydefs\fR(8), \fBttyadm\fR(8),
\fBpam\fR(3PAM), \fBlogindevperm\fR(4), \fBpam.conf\fR(4), \fButmpx\fR(4),
\fBattributes\fR(5), \fBenviron\fR(5), \fBpam_authtok_check\fR(5),
\fBpam_authtok_get\fR(5), \fBpam_authtok_store\fR(5), \fBpam_dhkeys\fR(5),
\fBpam_passwd_auth\fR(5), \fBpam_unix_account\fR(5), \fBpam_unix_auth\fR(5),
\fBpam_unix_session\fR(5), \fBsmf\fR(5)
.sp
.LP
\fI\fR
.SH NOTES
.sp
.LP
If a port is monitored by more than one \fBttymon\fR, it is possible for the
\fBttymon\fRs to send out prompt messages in such a way that they compete for
input.
.sp
.LP
The \fBpam_unix\fR(5) module is no longer supported. Similar functionality is
provided by \fBpam_authtok_check\fR(5), \fBpam_authtok_get\fR(5),
\fBpam_authtok_store\fR(5), \fBpam_dhkeys\fR(5), \fBpam_passwd_auth\fR(5),
\fBpam_unix_account\fR(5), \fBpam_unix_auth\fR(5), and
\fBpam_unix_session\fR(5).


usr/src/cmd/ttymon/ttymon.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

#include <stdio_ext.h>
#include <stdlib.h>
#include <fcntl.h>
#include <errno.h>
#include <poll.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/stropts.h>
#include <sys/resource.h>
#include <sys/termios.h>
#include <pwd.h>
#include <grp.h>
#include <unistd.h>
#include <ulimit.h>

#include "sac.h"
#include "ttymon.h"
#include "tmstruct.h"
#include "tmextern.h"

static	int	Initialized;

extern	int	Retry;
extern	struct	pollfd	*Pollp;
static	void	initialize();
static	void	open_all();
static	int	set_poll();
static	int	check_spawnlimit();
static	int	mod_ttydefs();

void	open_device();
void	set_softcar();

extern	int	check_session();
extern	void	sigalarm();
extern	void	revokedevaccess(char *, uid_t, gid_t, mode_t);
/* can't include libdevinfo.h */
extern int di_devperm_logout(const char *);

/*
 * 	ttymon	- a port monitor under SAC
 *		- monitor ports, set terminal modes, baud rate
 *		  and line discipline for the port
 *		- invoke service on port if connection request received
 *		- Usage: ttymon
 *			 ttymon -g [options]
 *			 Valid options are
 *			 -h
 *			 -d device
 *			 -l ttylabel
 *			 -t timeout
 *			 -m modules
 *			 -p prompt
 *
 *		- ttymon without args is invoked by SAC
 *		- ttymon -g is invoked by process that needs to
 *		  have login service on the fly
 */

int
main(int argc, char *argv[])
{
	int	nfds;
	extern	char	*lastname();

	/*
	 * Only the superuser should execute this command.
	 */
	if (getuid() != 0)
		return (1);

	/* remember original signal mask and dispositions */
	(void) sigprocmask(SIG_SETMASK, NULL, &Origmask);
	(void) sigaction(SIGINT, NULL, &Sigint);
	(void) sigaction(SIGALRM, NULL, &Sigalrm);
	(void) sigaction(SIGPOLL, NULL, &Sigpoll);
	(void) sigaction(SIGQUIT, NULL, &Sigquit);
	(void) sigaction(SIGCLD, NULL, &Sigcld);
	(void) sigaction(SIGTERM, NULL, &Sigterm);
#ifdef	DEBUG
	(void) sigaction(SIGUSR1, NULL, &Sigusr1);
	(void) sigaction(SIGUSR2, NULL, &Sigusr2);
#endif

	/*
	 * SIGQUIT needs to be ignored. Otherwise, hitting ^\ from
	 * console kills ttymon.
	 */
	(void) signal(SIGQUIT, SIG_IGN);

	if ((argc > 1) || (strcmp(lastname(argv[0]), "getty") == 0)) {
		ttymon_express(argc, argv);
		return (1);	/*NOTREACHED*/
	}

	initialize();

	for (;;) {
		nfds = set_poll(Pollp);
		if (!Reread_flag) {
			if (nfds > 0)
				do_poll(Pollp, nfds);
			else
				(void) pause();
		}
		/*
		 * READDB messages may arrive during poll or pause.
		 * So the flag needs to be checked again.
		 */
		if (Reread_flag) {
			Reread_flag = FALSE;
			re_read();
		}
		while (Retry) {
			Retry = FALSE;
			open_all();
		}
	}
}

static	void
initialize()
{
	struct	pmtab	*tp;
	register struct passwd *pwdp;
	register struct	group	*gp;
	struct	rlimit rlimit;
	extern	struct	rlimit	Rlimit;
	extern	 uid_t	Uucp_uid;
	extern	 gid_t	Tty_gid;

#ifdef 	DEBUG
	extern	opendebug();
#endif
	Initialized = FALSE;
	/*
	 * get_environ() must be called first,
	 * otherwise we don't know where the log file is
	 */
	get_environ();
	openttymonlog();
	openpid();
	openpipes();
	setup_PCpipe();

	log("PMTAG: %s", Tag);
	log("Starting state: %s",
	    (State == PM_ENABLED) ? "enabled" : "disabled");

#ifdef 	DEBUG
	opendebug(FALSE);
	debug("***** ttymon in initialize *****");
	log("debug mode is \t on");
#endif

	catch_signals();

	/* register to receive SIGPOLL when data comes to pmpipe */
	if (ioctl(Pfd, I_SETSIG, S_INPUT) < 0)
		fatal("I_SETSIG on pmpipe failed: %s", strerror(errno));

	sacpoll(); /* this is needed because there may be data already */

	Maxfiles = (int)ulimit(4, 0L);	/* get max number of open files */
	if (Maxfiles < 0)
		fatal("ulimit(4,0L) failed: %s", strerror(errno));

	if (getrlimit(RLIMIT_NOFILE, &Rlimit) == -1)
		fatal("getrlimit failed: %s", strerror(errno));

	rlimit.rlim_cur = rlimit.rlim_max = Rlimit.rlim_max;
	if (setrlimit(RLIMIT_NOFILE, &rlimit) == -1)
		fatal("setrlimit failed: %s", strerror(errno));

	(void) enable_extended_FILE_stdio(-1, -1);

	Maxfiles = rlimit.rlim_cur;
	Maxfds = Maxfiles - FILE_RESERVED;

	log("max open files = %d", Maxfiles);
	log("max ports ttymon can monitor = %d", Maxfds);

	read_pmtab();

	/*
	 * setup poll array
	 * 	- we allocate 10 extra pollfd so that
	 *	  we do not have to re-malloc when there is
	 *	  minor fluctuation in Nentries
	 */
	Npollfd = Nentries + 10;
	if (Npollfd > Maxfds)
		Npollfd = Maxfds;
	if ((Pollp = (struct pollfd *)
	    malloc((unsigned)(Npollfd * sizeof (struct pollfd))))
	    == NULL)
		fatal("malloc for Pollp failed");

	(void) mod_ttydefs();	/* just to initialize Mtime */
	if (check_version(TTYDEFS_VERS, TTYDEFS) != 0)
		fatal("check /etc/ttydefs version failed");

	read_ttydefs(NULL, FALSE);

	/* initialize global variables, Uucp_uid & Tty_gid */
	if ((pwdp = getpwnam(UUCP)) != NULL)
		Uucp_uid = pwdp->pw_uid;
	if ((gp = getgrnam(TTY)) == NULL)
		log("no group entry for <tty>, default is used");
	else
		Tty_gid = gp->gr_gid;
	endgrent();
	endpwent();
#ifdef	DEBUG
	debug("Uucp_uid = %u, Tty_gid = %u", Uucp_uid, Tty_gid);
#endif

	log("Initialization Completed");

	/* open the devices ttymon monitors */
	Retry = TRUE;
	while (Retry) {
		Retry = FALSE;
		for (tp = PMtab; tp; tp = tp->p_next) {
			if ((tp->p_status > 0) && (tp->p_fd == 0) &&
			    (tp->p_pid == 0) && !(tp->p_ttyflags & I_FLAG) &&
			    (!((State == PM_DISABLED) &&
			    ((tp->p_dmsg == NULL)||(*(tp->p_dmsg) == '\0'))))) {
				open_device(tp);
				if (tp->p_fd > 0)
					got_carrier(tp);
			}
		}
	}
	Initialized = TRUE;
}

static	void	free_defs();

/*
 *	open_all - open devices in pmtab if the entry is
 *	         - valid, fd = 0, and pid = 0
 */
static void
open_all()
{
	struct	pmtab	*tp;
	int	check_modtime;
	sigset_t cset;
	sigset_t tset;

#ifdef	DEBUG
	debug("in open_all");
#endif
	check_modtime = TRUE;

	for (tp = PMtab; tp; tp = tp->p_next) {
		if ((tp->p_status > 0) && (tp->p_fd == 0) &&
		    (tp->p_pid == 0) &&
		    !(tp->p_ttyflags & I_FLAG) && (!((State == PM_DISABLED) &&
		    ((tp->p_dmsg == NULL)||(*(tp->p_dmsg) == '\0'))))) {
			/*
			 * if we have not check modification time and
			 * /etc/ttydefs was modified, need to re-read it
			 */
			if (check_modtime && mod_ttydefs()) {
				check_modtime = FALSE;
				(void) sigprocmask(SIG_SETMASK, NULL, &cset);
				tset = cset;
				(void) sigaddset(&tset, SIGCLD);
				(void) sigprocmask(SIG_SETMASK, &tset, NULL);
				free_defs();
#ifdef	DEBUG
				debug("/etc/ttydefs is modified, re-read it");
#endif
				read_ttydefs(NULL, FALSE);
				(void) sigprocmask(SIG_SETMASK, &cset, NULL);
			}
			open_device(tp);
			if (tp->p_fd > 0)
				got_carrier(tp);
		} else if (((tp->p_status == LOCKED) ||
		    (tp->p_status == SESSION) ||
		    (tp->p_status == UNACCESS)) &&
		    (tp->p_fd > 0) &&
		    (!((State == PM_DISABLED) &&
		    ((tp->p_dmsg == NULL)||(*(tp->p_dmsg) == '\0'))))) {
			if (check_modtime && mod_ttydefs()) {
				check_modtime = FALSE;
				(void) sigprocmask(SIG_SETMASK, NULL, &cset);
				tset = cset;
				(void) sigaddset(&tset, SIGCLD);
				(void) sigprocmask(SIG_SETMASK, &tset, NULL);
				free_defs();
#ifdef	DEBUG
				debug("/etc/ttydefs is modified, re-read it");
#endif
				read_ttydefs(NULL, FALSE);
				(void) sigprocmask(SIG_SETMASK, &cset, NULL);
			}
			tp->p_status = VALID;
			open_device(tp);
			if (tp->p_fd > 0)
				got_carrier(tp);
		}
	}
}

void
set_softcar(pmptr)
struct	pmtab	*pmptr;
{

	int fd, val = 0;

#ifdef	DEBUG
	debug("in set_softcar");
#endif
	/*
	 * If soft carrier is not set one way or
	 * the other, leave it alone.
	 */
	if (*pmptr->p_softcar == '\0')
		return;

	if (*pmptr->p_softcar == 'y')
		val = 1;

	if ((fd = open(pmptr->p_device, O_RDONLY|O_NONBLOCK|O_NOCTTY)) < 0) {
		log("open (%s) failed: %s", pmptr->p_device, strerror(errno));
		return;
	}

	if (ioctl(fd, TIOCSSOFTCAR, &val) < 0)
		log("set soft-carrier (%s) failed: %s", pmptr->p_device,
		    strerror(errno));

	close(fd);
}


/*
 *	open_device(pmptr)	- open the device
 *				- check device lock
 *				- change owner of device
 *				- push line disciplines
 *				- set termio
 */

void
open_device(pmptr)
struct	pmtab	*pmptr;
{
	int	fd, tmpfd;
	struct	sigaction	sigact;

#ifdef	DEBUG
	debug("in open_device");
#endif

	if (pmptr->p_status == GETTY) {
		revokedevaccess(pmptr->p_device, 0, 0, 0);

		if ((fd = open(pmptr->p_device, O_RDWR)) == -1)
			fatal("open (%s) failed: %s", pmptr->p_device,
			    strerror(errno));

	} else {
		if (check_spawnlimit(pmptr) == -1) {
			pmptr->p_status = NOTVALID;
			log("service <%s> is respawning too rapidly",
			    pmptr->p_tag);
			return;
		}
		if (pmptr->p_fd > 0) { /* file already open */
			fd = pmptr->p_fd;
			pmptr->p_fd = 0;
		} else if ((fd = open(pmptr->p_device, O_RDWR|O_NONBLOCK))
		    == -1) {
			log("open (%s) failed: %s", pmptr->p_device,
			    strerror(errno));
			if ((errno ==  ENODEV) || (errno == EBUSY)) {
				pmptr->p_status = UNACCESS;
				Nlocked++;
				if (Nlocked == 1) {
				    sigact.sa_flags = 0;
				    sigact.sa_handler = sigalarm;
				    (void) sigemptyset(&sigact.sa_mask);
				    (void) sigaction(SIGALRM, &sigact, NULL);
				    (void) alarm(ALARMTIME);
				}
			} else
				Retry = TRUE;
			return;
		}
		/* set close-on-exec flag */
		if (fcntl(fd, F_SETFD, 1) == -1)
			fatal("F_SETFD fcntl failed: %s", strerror(errno));

		if (tm_checklock(fd) != 0) {
			pmptr->p_status = LOCKED;
			(void) close(fd);
			Nlocked++;
			if (Nlocked == 1) {
				sigact.sa_flags = 0;
				sigact.sa_handler = sigalarm;
				(void) sigemptyset(&sigact.sa_mask);
				(void) sigaction(SIGALRM, &sigact, NULL);
				(void) alarm(ALARMTIME);
			}
			return;
		}
		if (check_session(fd) != 0) {
			if ((Initialized) && (pmptr->p_inservice != SESSION)) {
				log("Warning -- active session exists on <%s>",
				    pmptr->p_device);
			} else {
				/*
				 * this may happen if a service is running
				 * and ttymon dies and is restarted,
				 * or another process is running on the
				 * port.
				 */
				pmptr->p_status = SESSION;
				pmptr->p_inservice = 0;
				(void) close(fd);
				Nlocked++;
				if (Nlocked == 1) {
					sigact.sa_flags = 0;
					sigact.sa_handler = sigalarm;
					(void) sigemptyset(&sigact.sa_mask);
					(void) sigaction(SIGALRM, &sigact,
					    NULL);
					(void) alarm(ALARMTIME);
				}
				return;
			}
		}
		pmptr->p_inservice = 0;
	}

	if (pmptr->p_ttyflags & H_FLAG) {
		/* drop DTR */
		(void) hang_up_line(fd);
		/*
		 * After hang_up_line, the stream is in STRHUP state.
		 * We need to do another open to reinitialize streams
		 * then we can close one fd
		 */
		if ((tmpfd = open(pmptr->p_device, O_RDWR|O_NONBLOCK)) == -1) {
			log("open (%s) failed: %s", pmptr->p_device,
			    strerror(errno));
			Retry = TRUE;
			(void) close(fd);
			return;
		}
		(void) close(tmpfd);
	}

#ifdef DEBUG
	debug("open_device (%s), fd = %d", pmptr->p_device, fd);
#endif

	/* Change ownership of the tty line to root/uucp and */
	/* set protections to only allow root/uucp to read the line. */

	if (pmptr->p_ttyflags & (B_FLAG|C_FLAG))
		(void) fchown(fd, Uucp_uid, Tty_gid);
	else
		(void) fchown(fd, ROOTUID, Tty_gid);
	(void) fchmod(fd, 0620);

	if ((pmptr->p_modules != NULL)&&(*(pmptr->p_modules) != '\0')) {
		if (push_linedisc(fd, pmptr->p_modules, pmptr->p_device)
		    == -1) {
			Retry = TRUE;
			(void) close(fd);
			return;
		}
	}

	if (initial_termio(fd, pmptr) == -1)  {
		Retry = TRUE;
		(void) close(fd);
		return;
	}

	di_devperm_logout((const char *)pmptr->p_device);
	pmptr->p_fd = fd;
}

/*
 *	set_poll(fdp)	- put all fd's in a pollfd array
 *			- set poll event to POLLIN and POLLMSG
 *			- return number of fd to be polled
 */

static	int
set_poll(fdp)
struct pollfd *fdp;
{
	struct	pmtab	*tp;
	int 	nfd = 0;

	for (tp = PMtab; tp; tp = tp->p_next) {
		if (tp->p_fd > 0)  {
			fdp->fd = tp->p_fd;
			fdp->events = POLLIN;
			fdp++;
			nfd++;
		}
	}
	return (nfd);
}

/*
 *	check_spawnlimit	- return 0 if spawnlimit is not reached
 *				- otherwise return -1
 */
static	int
check_spawnlimit(pmptr)
struct	pmtab	*pmptr;
{
	time_t	now;

	(void) time(&now);
	if (pmptr->p_time == 0L)
		pmptr->p_time = now;
	if (pmptr->p_respawn >= SPAWN_LIMIT) {
		if ((now - pmptr->p_time) < SPAWN_INTERVAL) {
			pmptr->p_time = now;
			pmptr->p_respawn = 0;
			return (-1);
		}
		pmptr->p_time = now;
		pmptr->p_respawn = 0;
	}
	pmptr->p_respawn++;
	return (0);
}

/*
 * mod_ttydefs	- to check if /etc/ttydefs has been modified
 *		- return TRUE if file modified
 *		- otherwise, return FALSE
 */
static	int
mod_ttydefs()
{
	struct	stat	statbuf;
	extern	long	Mtime;
	if (stat(TTYDEFS, &statbuf) == -1) {
		/* if stat failed, don't bother reread ttydefs */
		return (FALSE);
	}
	if ((long)statbuf.st_mtime != Mtime) {
		Mtime = (long)statbuf.st_mtime;
		return (TRUE);
	}
	return (FALSE);
}

/*
 *	free_defs - free the Gdef table
 */
static	void
free_defs()
{
	int	i;
	struct	Gdef	*tp;
	tp = &Gdef[0];
	for (i = 0; i < Ndefs; i++, tp++) {
		free(tp->g_id);
		free(tp->g_iflags);
		free(tp->g_fflags);
		free(tp->g_nextid);
		tp->g_id = NULL;
		tp->g_iflags = NULL;
		tp->g_fflags = NULL;
		tp->g_nextid = NULL;
	}
	Ndefs = 0;
}

/*
 * struct Gdef *get_speed(ttylabel)
 *	- search "/etc/ttydefs" for speed and term. specification
 *	  using "ttylabel". If "ttylabel" is NULL, default
 *	  to DEFAULT
 * arg:	  ttylabel - label/id of speed settings.
 */

struct Gdef *
get_speed(char *ttylabel)
{
	register struct Gdef *sp;
	extern   struct Gdef DEFAULT;

	if ((ttylabel != NULL) && (*ttylabel != '\0')) {
		if ((sp = find_def(ttylabel)) == NULL) {
			log("unable to find <%s> in \"%s\"", ttylabel, TTYDEFS);
			sp = &DEFAULT; /* use default */
		}
	} else sp = &DEFAULT; /* use default */
	return (sp);
}

/*
 * setup_PCpipe()	- setup the pipe between Parent and Children
 *			- the pipe is used for a tmchild to send its
 *			  pid to inform ttymon that it is about to
 *			  invoke service
 *			- the pipe also serves as a mean for tmchild
 *			  to detect failure of ttymon
 */
void
setup_PCpipe()
{
	int	flag = 0;

	if (pipe(PCpipe) == -1)
		fatal("pipe() failed: %s", strerror(errno));

	/* set close-on-exec flag */
	if (fcntl(PCpipe[0], F_SETFD, 1) == -1)
		fatal("F_SETFD fcntl failed: %s", strerror(errno));

	if (fcntl(PCpipe[1], F_SETFD, 1) == -1)
		fatal("F_SETFD fcntl failed: %s", strerror(errno));

	/* set O_NONBLOCK flag */
	if (fcntl(PCpipe[0], F_GETFL, flag) == -1)
		fatal("F_GETFL failed: %s", strerror(errno));

	flag |= O_NONBLOCK;
	if (fcntl(PCpipe[0], F_SETFL, flag) == -1)
		fatal("F_SETFL failed: %s", strerror(errno));

	/* set message discard mode */
	if (ioctl(PCpipe[0], I_SRDOPT, RMSGD) == -1)
		fatal("I_SRDOPT RMSGD failed: %s", strerror(errno));

	/* register to receive SIGPOLL when data come */
	if (ioctl(PCpipe[0], I_SETSIG, S_INPUT) == -1)
		fatal("I_SETSIG S_INPUT failed: %s", strerror(errno));

#ifdef 	DEBUG
	log("PCpipe[0]\t = %d", PCpipe[0]);
	log("PCpipe[1]\t = %d", PCpipe[1]);
#endif
}
